keps/sig-network/4559-kubelet-probes/README.md

```diff
@@ -0,0 +1,1037 @@
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### API Changes to Fix the SSRF
+
+Regardless of what else we do, we will deprecate the
+`TCPSocketAction.Host` and `HTTPGetAction.Host` fields, and allow them
+to be blocked by Pod Security admission.
+
+Beyond that, there are two (and a half) possible plans:
+
+1. Change the semantics of the existing `TCPSocketAction`,
+   `HTTPGetAction`, and `GRPCAction` probe types and say that
+   henceforth the probes will be done in the pod network namespace
+   rather than being sent over the pod network (unless you use the
+   deprecated `Host` field to request an alternate probe endpoint; we
+   probably shouldn't just move those kinds of probes into the pod
+   network namespace since that might require a "hairpin" connection
+   to work when it wasn't required before).
+
+2. Leave the existing probe types with their existing semantics, but
+   deprecate them and add new probe types with the
+   inside-the-network-namespace semantics. (`v1.ProbeHandler`
+   currently has `httpGet` and `tcpSocket` fields; these could
+   potentially become just `http` and `tcp`. It's less obvious what
+   `grpc` could be renamed to.)
+
+2½. Deprecate-and-replace `httpGet` and `tcpSocket` as in (2), but
+    change the semantics of `grpc` as in (1), since GRPC probes are
+    already more constrained than the TCP and HTTP probes anyway (and
+    I don't have a good suggestion for renaming them).
+
+### Kubelet/Runtime Changes to Fix the NetworkPolicy Hole
+
+Again, a few possible plans. Some of these ideas are pretty bad, but
+may have good ideas within them that could be remixed into other
+plans...
+
+1. Rather than performing pod probes from the node IP, each kubelet
+   could launch a "probing pod", and perform probes from there. A
+   created-by-default AdminNetworkPolicy could force all pods to
+   accept connections from the kubelet probe pods. This is perhaps
+   _slightly_ better than the current situation since allowing
+   connections from the probe pods would not allow any unintended
+   additional traffic from anyone else the way that allowing
+   connections from the node IP does.
+
+     - The simplest / most portable way to do this would be to have
+       the "probing pod" be an ordinary pod-network pod, though (given
+       current AdminNetworkPolicy rules) this would means that all
+       pods on all nodes would have to accept connections from all
+       probing pods, meaning an attacker who was able to compromise
+       any probing pod would then be able to connect to any pod in the
+       cluster.
+
+     - Alternatively, the "probing pod" could be a static pod with a
```

Comment:
Yeah, we've been talking about "pods attached only to a secondary network, not the cluster default pod network" in OpenShift, and one tricky bit is: how does DNS work for them? The current thinking is that pods that "aren't attached to the cluster default pod network" will still actually be attached to the cluster default pod network, just with routing/policy set up to only allow it to be used in very limited ways (like DNS).

But having a special "infra network" attached to every pod would be another way of doing that same thing...

The trick is finding a CIDR range to use; we can't use `169.254.0.0/16` because many existing clusters expect pods to be able to connect to the `169.254.169.254` outside the node. IPv6 link-local addresses would work, but we can't make the infra network be IPv6-only because then we couldn't do health checks for IPv4-only pods over it.

It's still an interesting idea...

Reply:
aslo link-local addresses are commonly used by different things to provide this sort of mechanism, if following that path we should make clear what is the range reserved for kubernetes probes ... 

 IPv6 has this concept of address scopes https://datatracker.ietf.org/doc/html/rfc4007#section-6 that solves this problem, as we can create the probe address scope range ... if we just could ensure IPv6 is enable on all clusters, adding this IPv6 capability to pods will certainly solve this problem with very low effort